### ISSUE TITLE
Undertow/Invalid argument with no request body 

### ISSUE THREAD

**Issue by fmance** [palantir/conjure-java#317]
## What happened?
- have an undertow endpoint with an optional request body (i.e., `Optional<BodyType>`)
- calling the endpoint without passing any request body throws an exception

```
INFO  [2019-04-03T11:01:14.46Z]  com.palantir.conjure.java.undertow.runtime.ConjureExceptionHandler: Error handling request (errorInstanceId: 2e7227fe-9233-4bab-995f-f86d937aee07, errorName: Default:InvalidArgument, throwable1_type: map[actualTypeArguments:[BodyType] rawType:java.util.Optional typeName:java.util.Optional<BodyType>]) (throwable2_message: No content to map due to end-of-input
 at [Source: (io.undertow.io.UndertowInputStream); line: 1, column: 0])
com.palantir.conjure.java.api.errors.ServiceException: ServiceException: INVALID_ARGUMENT (Default:InvalidArgument)
	at com.palantir.conjure.java.undertow.runtime.ConjureExceptionHandler.frameworkException(ConjureExceptionHandler.java:140)
	at com.palantir.conjure.java.undertow.runtime.ConjureExceptionHandler.handleRequest(ConjureExceptionHandler.java:74)
	at com.palantir.tracing.undertow.TracedOperationHandler.handleRequest(TracedOperationHandler.java:68)
	at com.palantir.conjure.java.undertow.runtime.LoggingContextHandler.handleRequest(LoggingContextHandler.java:40)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:364)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1871)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1451)
	at java.lang.Thread.run(Thread.java:834)
Caused by: com.palantir.conjure.java.undertow.runtime.FrameworkException: Failed to deserialize response stream. Syntax error?
	at com.palantir.conjure.java.undertow.runtime.FrameworkException.unprocessableEntity(FrameworkException.java:40)
	at com.palantir.conjure.java.undertow.runtime.Encodings$AbstractJacksonEncoding.lambda$deserializer$1(Encodings.java:67)
	at com.palantir.conjure.java.undertow.runtime.ConjureBodySerDe$EncodingDeserializerRegistry.deserialize(ConjureBodySerDe.java:148)

......


Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: No content to map due to end-of-input
 at [Source: (io.undertow.io.UndertowInputStream); line: 1, column: 0]
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:59)
	at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1342)
	at com.fasterxml.jackson.databind.ObjectReader._initForReading(ObjectReader.java:358)
	at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:1593)
	at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1185)
	at com.palantir.conjure.java.undertow.runtime.Encodings$AbstractJacksonEncoding.lambda$deserializer$1(Encodings.java:62)
	at com.palantir.conjure.java.undertow.runtime.ConjureBodySerDe$EncodingDeserializerRegistry.deserialize(ConjureBodySerDe.java:148)
```

## What did you want to happen?
When using Jersey instead of undertow, the body is mapped to `null`, so undertow should preserve the same behavior.

The error seems to be coming from this line https://github.com/palantir/conjure-java/blob/develop/conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/Encodings.java#L62

**Comment by fmance** [comment]
cc @iamdanfox

**Comment by carterkozak** [comment]
Oops, I'll put something together later today.
Does jersey map this properly to an empty optional, or does it provide a `null` optional?

**Comment by carterkozak** [comment]
conjure-undertow works for mapping bodies with `application/json` string `null` into an empty optional, but it doesn't handle the zero-length body case.

**Comment by fmance** [comment]
@cakofony Jersey maps it to `null` (i.e., a `null` is received in the "resource" implementing the service). So for consistency, I guess it should be `null` in undertow as well, even though `Optional.empty()` would seem more intuitive ?

**Comment by carterkozak** [comment]
I would consider receiving a null optional a bug in the jersey implementation. I wouldn’t consider fixing this properly in conjure-undertow as a break because the jersey implementation must already support all three inputs: null (no body), empty optional (body with json null), and present optional.

**Comment by fmance** [comment]
Makes sense 👍

**Comment by carterkozak** [comment]
fixed by #319
